Авто-логирование

ASherstobitov opened the pull request.

Commit:
Авто-логирование

Commit:
Авто-логирование

petrelevich commented on hw-autoLogger/src/main/java/ru/otus/proxy/CalculatorInterface.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface CalculatorInterface {
+
+    @Log
```
аннотация должна быть только на имплементации.
на интерфейсе быть не должно.

petrelevich commented on hw-autoLogger/src/main/java/ru/otus/proxy/MyProxy.java:
```diff
@@ -0,0 +1,57 @@
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+            Arrays.stream(calculatorInterface.getClass().getMethods())
```
делать рефлексивный разбор при каждом вызове метода не эффективно.
Это надо сделать один раз в конструкторе и потом переиспользовать.

Commit:
Авто-логирование
Примечания:
-Вынес рефлексивный разбор в конструктор
-Убрал аннотацию из интерфейса

petrelevich commented on hw-autoLogger/src/main/java/ru/otus/proxy/MyProxy.java:
```diff
@@ -0,0 +1,63 @@
+public class MyProxy {
+
+
+    private static List<Method> methods;
```
лучше Set

petrelevich commented on hw-autoLogger/src/main/java/ru/otus/proxy/MyProxy.java:
```diff
@@ -0,0 +1,63 @@
+
+
+            methods = Arrays.stream(calculatorInterface.getClass().getMethods())
+                    .filter(e -> Arrays.stream(e.getDeclaredAnnotations()).anyMatch(a -> a.annotationType().equals(Log.class)))
```
посмотрите метод:
isAnnotationPresent

petrelevich commented on hw-autoLogger/src/main/java/ru/otus/proxy/MyProxy.java:
```diff
@@ -0,0 +1,63 @@
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+
+            methods.stream()
```
эффективнее в methods положить не метод, а метод+сигнатура, чтобы тут можно было вызвать просто contains.
т.е. нужен метод, который сделает: method.getName() + method.getParameterTypes()
этот метод будет использован в конструкторе для заполнения коллекции и потом в invoke для проверки наличия нужного метода в коллекции.
